### DISCUSSION TITLE
tsdl: a tool to download and build parsers from remote git

### DISCUSSION THREAD

**Discussion by stackmystack** [tree-sitter/tree-sitter#3629] [Libraries & tools]
Hello,

I'm not sure if there's a tool that already exists doing the same thing, but I developed [tsdl](https://github.com/stackmystack/tsdl) for my own personal needs, and I'm now happy to share it with the tree-sitter community since it replaced so many ad-hoc shell scripts I had for many different contexts, and my latest integration with [ruby-tree-sitter](https://github.com/Faveod/ruby-tree-sitter) was a breeze.

`tsdl` reads a `toml` config to specify the list of parsers you want, and downloads, and builds parsers using `tree-sitter-cli`.

I've written it in rust, with the specific purpose of getting myself more familiar with rust, so obviously there's a lot of things that could be improved (and I have a list I'm working on little by little), or changed (e.g. do I really need `tokio` when I really want a simple threadpool? It was fun to get to know async rust anyway).

Your feedback is most welcome.

**Comment by sogaiu** [comment]
I recently put together [ts-playground-setup](https://github.com/sogaiu/ts-playground-setup) which functionally has some similarities as it uses the `tree-sitter` cli to accomplish certain tasks.  Its main purpose is to allow one to generate a directory which can be hosted somewhere (locally too ofc) containing a slight "remix" of the playground so that one can choose to have multiple grammars [1] to select from (like the "official" hosted playground).  The result can look like [this](https://sogaiu.github.io/ts-playground-setup/playground.html).

It doesn't build shared objects, but it does generate `parser.c` and friends as a prerequisite to then generating grammar `.wasm` files.  It also builds `tree-sitter.js` and `tree-sitter.wasm`, clones some grammar repositories, fetches some `.js` / `.css`, patches `playground.html`, etc.

One thing I've been wary of are changes in behavior / API of the `tree-sitter` cli.  I wanted this tool to work with different versions (but also different commits) of the cli, especially when investigating issues that might have to do with version differences.  I think the tree-sitter maintainers are way overburdened though and my impression is that having a stable API of the cli (not the library) is, relatively speaking, not as important as other concerns.  Just mentioning this here as something to be wary of when trying to build on top of the cli.  It's not the same kind of thing as trying to build on top of `git` -- which has way more resources dedicated toward it and AFAICT has a history of (and possibly a policy of) trying to maintain compatibility (which can lead to weird default behaviors...).

---

[1] I went for a plain text file -- not as fancy as your `.toml` choice :)

**Reply by stackmystack** [reply]
Interesting.

>  I wanted this tool to work with different versions (but also different commits) of the cli

I don't have such requirements, and I don't know how the maintainers regard releases, but I think I would never want `tsdl` to work with non-releases `tree-sitter-cli`. That being said, if you specify `--tree-sitter-version="commit-sha1"`, `tsdl` will figure out which release it should fetch from github, and use that `cli`.

> my impression is that having a stable API of the cli (not the library) is, relatively speaking, not as important as other concerns.
> Just mentioning this here as something to be wary of when trying to build on top of the cli. 

Honestly I get it. I'm not even bothered by it because I maintain the ruby bindings for tree-sitter, and all I care about is the latest tree-sitter, so I will make sure that the latest `tsdl` follows the latest cli API. What I do care about though, is having compatibility with pinned parsers. That being said, too, you can with tsdl compile for older tree-sitter versions where the `cli` doesn't provide the `build` command, by providing the `cmd` value for parsers, and put `make` in it, just like `typescript` in the example.

Maybe I can provide a `tsdl playground` at some point that serves a playground of the configured parsers (building them to wasm ofc). I haven't looked into your project, but if it's easy to integrate maybe I can pull from it.

**Reply by sogaiu** [reply]
> That being said, if you specify --tree-sitter-version="commit-sha1", tsdl will figure out which release it should fetch from github, and use that cli.

Yeah I saw that.  Nice!

> Maybe I can provide a tsdl playground at some point that serves a playground of the configured parsers (building them to wasm ofc).

Right, if one had a use for `tsdl`, then a playground feature might be a natural desire :)

> I haven't looked into your project, but if it's easy to integrate maybe I can pull from it.

As I didn't use Rust, likely the code won't be of much use, but possibly the steps taken [1] and/or some of the description [here](https://github.com/sogaiu/ts-playground-setup/blob/9efba10f2fe10eec14ef13534fcb912c62b37367/README.md#implementation) might be worth glancing at.

---

[1] If you go for building `.wasm` bits and opt to use emsdk, may be you'll find [this comment from an emsdk person](https://github.com/emscripten-core/emsdk/issues/1142#issuecomment-1334065131) to be handy.
